### PR TITLE
Esapi r138 and minor fixed

### PR DESCRIPTION
**Pull request by AndreasFuchsSIT**
Add support for TPM2.0 r1.38 functions and some minor fixes.

### CODE REVIEW

File: test/integration/esys-hierarchychangeauth.int.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    goto_if_error(r, "Error: TR_SetAuth", error);
+
+    ESYS_TR primaryHandle_handle;
+    RSRC_NODE_T *primaryHandle_node;
+    TPM2B_PUBLIC *outPublic;
+    TPM2B_CREATION_DATA *creationData;
+    TPM2B_DIGEST *creationHash;
+    TPMT_TK_CREATION *creationTicket;
+
+    r = Esys_CreatePrimary(esys_context, ESYS_TR_RH_OWNER, ESYS_TR_PASSWORD,
```
Comment (tstruk) [inline test/integration/esys-hierarchychangeauth.int.c]: Please add a comments describing the usecase i.e. what we want to test and why. It makes it easier to follow,

File: test/integration/esys-hierarchychangeauth.int.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                           ESYS_TR_NONE, ESYS_TR_NONE, &inSensitivePrimary, &inPublic,
+                           &outsideInfo, &creationPCR, &primaryHandle_handle,
+                           &outPublic, &creationData, &creationHash,
+                           &creationTicket);
+    goto_if_error(r, "Error esys create primary", error);
+
+    r = Esys_TR_SetAuth(esys_context, ESYS_TR_RH_OWNER, &newAuth);
+    goto_if_error(r, "Error SetAuth", error);
+
+    r = Esys_CreatePrimary(esys_context, ESYS_TR_RH_OWNER, ESYS_TR_PASSWORD,
```
Comment (tstruk) [inline test/integration/esys-hierarchychangeauth.int.c]: Use flush context to cleanup after the test.

File: esapi/esapi/Esys_PolicyTemplate.c [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    r = Tss2_Sys_PolicyTemplate_Complete(esysContext->sys);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Error finish (ExecuteFinish) PolicyTemplate: %" PRIx32, r);
+        esysContext->state = _ESYS_STATE_ERRORRESPONSE;
+        return r;;
+    }
+    esysContext->state = _ESYS_STATE_FINISHED;
+
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_PolicyTemplate.c]: Add new line

File: esapi/esapi/Esys_CreateLoaded.c [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+    return r;
+
+error_cleanup:
+    Esys_TR_Close(esysContext, objectHandle);
+    if (outPrivate != NULL)
+        SAFE_FREE(*outPrivate);
+    SAFE_FREE(loutPublic);
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_CreateLoaded.c]: Add new line

File: esapi/esapi/Esys_EncryptDecrypt2.c [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+    return r;
+
+error_cleanup:
+    if (outData != NULL)
+        SAFE_FREE(*outData);
+    if (ivOut != NULL)
+        SAFE_FREE(*ivOut);
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_EncryptDecrypt2.c]: Add new line

File: esapi/esapi/Esys_FieldUpgradeData.c [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+    return r;
+
+error_cleanup:
+    if (nextDigest != NULL)
+        SAFE_FREE(*nextDigest);
+    if (firstDigest != NULL)
+        SAFE_FREE(*firstDigest);
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_FieldUpgradeData.c]: Add new line

File: esapi/esapi/Esys_FieldUpgradeStart.c [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    r = Tss2_Sys_FieldUpgradeStart_Complete(esysContext->sys);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Error finish (ExecuteFinish) FieldUpgradeStart: %" PRIx32, r);
+        esysContext->state = _ESYS_STATE_ERRORRESPONSE;
+        return r;;
+    }
+    esysContext->state = _ESYS_STATE_FINISHED;
+
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_FieldUpgradeStart.c]: Add new line

File: esapi/esapi/Esys_FirmwareRead.c [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    esysContext->state = _ESYS_STATE_FINISHED;
+
+    return r;
+
+error_cleanup:
+    if (fuData != NULL)
+        SAFE_FREE(*fuData);
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_FirmwareRead.c]: Add new line

File: esapi/esapi/Esys_PolicyAuthorizeNV.c [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+    r = Tss2_Sys_PolicyAuthorizeNV_Complete(esysContext->sys);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Error finish (ExecuteFinish) PolicyAuthorizeNV: %" PRIx32, r);
+        esysContext->state = _ESYS_STATE_ERRORRESPONSE;
+        return r;;
+    }
+    esysContext->state = _ESYS_STATE_FINISHED;
+
+    return r;
+}
```
Comment (tstruk) [inline esapi/esapi/Esys_PolicyAuthorizeNV.c]: Add new line

File: esapi/esapi/Esys_NV_ReadLock.c [hunk 10]
```diff
@@ -264,6 +264,18 @@ Esys_NV_ReadLock_finish(
         esysContext->state = _ESYS_STATE_ERRORRESPONSE;
         return r;;
     }
+
+    ESYS_TR nvIndex = esysContext->in.NV_Write.nvIndex;
+        RSRC_NODE_T *nvIndexNode;
```
Comment (tstruk) [inline esapi/esapi/Esys_NV_ReadLock.c]: fix indent

File: esapi/esapi/Esys_NV_WriteLock.c [hunk 11]
```diff
@@ -264,6 +264,18 @@ Esys_NV_WriteLock_finish(
         esysContext->state = _ESYS_STATE_ERRORRESPONSE;
         return r;;
     }
+
+    ESYS_TR nvIndex = esysContext->in.NV_Write.nvIndex;
+        RSRC_NODE_T *nvIndexNode;
```
Comment (tstruk) [inline esapi/esapi/Esys_NV_WriteLock.c]: fix indent

File: test/integration/esys-hierarchychangeauth.int.c (hunk 1 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-hierarchychangeauth.int.c]: see line 40ff

File: test/integration/esys-hierarchychangeauth.int.c (hunk 2 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-hierarchychangeauth.int.c]: done

File: esapi/esapi/Esys_PolicyTemplate.c (hunk 3 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_PolicyTemplate.c]: done

File: esapi/esapi/Esys_CreateLoaded.c (hunk 4 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_CreateLoaded.c]: done

File: esapi/esapi/Esys_EncryptDecrypt2.c (hunk 5 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_EncryptDecrypt2.c]: done

File: esapi/esapi/Esys_FieldUpgradeData.c (hunk 6 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_FieldUpgradeData.c]: done

File: esapi/esapi/Esys_FieldUpgradeStart.c (hunk 7 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_FieldUpgradeStart.c]: done

File: esapi/esapi/Esys_FirmwareRead.c (hunk 8 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_FirmwareRead.c]: done

File: esapi/esapi/Esys_PolicyAuthorizeNV.c (hunk 9 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_PolicyAuthorizeNV.c]: done

File: esapi/esapi/Esys_NV_ReadLock.c (hunk 10 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_NV_ReadLock.c]: done

File: esapi/esapi/Esys_NV_WriteLock.c (hunk 11 above)
Comment (AndreasFuchsSIT) [reply esapi/esapi/Esys_NV_WriteLock.c]: done

**Pull request merged**